quipucords/scanner/normalizer.py

```diff
@@ -0,0 +1,229 @@
+                fact_name,
+                unexpected_keys,
+            )
+
+
+class FactMapper:
+    """
+    Utility for mapping raw facts to normalizers and validators.
+
+    Must be used in conjunction with Normalizer classes.
+    """
+
+    def __init__(
+        self,
+        raw_fact_key: str | list[str] | None,
+        normalizer: callable,
+        *,
+        validators=None,
+        dependencies: list[str] = None,
+    ) -> None:
+        """Instantiate FactMapper.
+
+        :param raw_fact_key: raw fact name or list of raw fact names used to normalize a
+            fact
+        :param normalizer: function used to normalize raw fact. When raw_fact_key is a
+            str, signature of normalizer SHOULD be fn(raw_fact_value, **dependencies)
+            where dependencies are kwargs from dependencies. When raw_fact_key is a
+            list, signature should be fn(**kwargs), where kwargs are raw facts and
+            dependencies.
+        :param validators: list of functions used to validate the output (signature
+            fn(fact_value) -> bool)
+        :param dependencies: list of normalized facts that this fact depends
+        """
+        if isinstance(raw_fact_key, str):
+            self.raw_fact_keys = [raw_fact_key]
+        elif raw_fact_key is None:
+            self.raw_fact_keys = []
+        else:
+            self.raw_fact_keys = raw_fact_key
+        self.fact_name = None
+        self.normalizer = normalizer
+        self.dependencies = dependencies or []
+        self.validators = validators or []
+
+    def __set_name__(self, normalizer, name):
+        """Configure FacMapper and Normalizer."""
+        if not issubclass(normalizer, BaseNormalizer):
+            raise AssertionError(
+                "FactMapper instances can only be bound to Normalizers"
+            )
+        if normalizer.fields == NotImplemented:
+            # concrete normalizer classes require class attribute 'fields' to map
+            # fact_name to FactMappers in the exact order they were created (see
+            # normalize method on BaseNormalizer and the method
+            # `_ensure_dependencies_exist` below).
+            # 'fields' dict can't be created at BaseNormalizer level otherwise all
+            # implementations will share it. Can't be done at normalizer instance level
+            # either because this method don't have access to it (we are dealing with
+            # class construction here).
+            # As odd as it seems to create something for another class here, given the
+            # symbiotic relationship between Normalizers and FactMappers, this seems to
+            # be the sweetspot for it and a good trade off.
+            normalizer.fields = {}
+        # ensure dependencies are created BEFORE this one
+        self._ensure_dependencies_exist(normalizer)
+        normalizer.fields[name] = self
+        self.fact_name = name
```

Comment:
how about this @mirekdlugosz ? got rid of the metaclass, but I still feel the wall of text might be helpful for future readers

Reply:
My gut instinct tells me `None` might be more appropriate than `NotImplemented`, but I've only very rarely actually used `NotImplemented`, and I don't know what side effects I may be missing from using one or the other here.

Otherwise I think I like this idea, and I am grateful for the wall of text comment!

Reply:
I never used not `NotImplemented` before but it felt adequate to this use case. Lemme check what happens if you don't add a fact mapper and try to call `.normalize`

Reply:
mmmm 

```
AttributeError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 n.normalize()

File ~/src/quipucords/quipucords/scanner/normalizer.py:60, in BaseNormalizer.normalize(self)
     58     return self._normalized
     59 self._normalized = {}
---> 60 for fact_name in self.fields.keys():
     61     norm_result = self._normalize_fact(fact_name)
     62     self._normalized[fact_name] = norm_result.value

AttributeError: 'NotImplementedType' object has no attribute 'keys'

```

Reply:
IDK, I'm on the fence. If u really prefer `None` I can gladly change it.

Reply:
replaced `NotImplemented` with `None`

Reply:
Late to the party, but:

- agreed that `None` is better than `NotImplemented`, which has pretty specific meaning for Python internals.
- nothing wrong with wall of text. We deal with relatively complex problem to begin with, it won't hurt to have a little bit of prose to better understand the rationale and possible pitfalls.